### PR TITLE
Ensure debug-mode apps are always attached on iOS.

### PR DESCRIPTION
**Pull request by sjindel-google**
This is the implementation of go/decommissioning-dbc-engine. The flag was is introduced is actually already being passed by Flutter tools for every debug-mode launch, but ignored in the Engine.

The tests are in the Flutter Framework PR https://github.com/flutter/flutter/pull/37043.

I'm working on the Wiki page mentioned in the warning message and will update that message shortly.

### CODE REVIEW

File: runtime/dart_vm.cc [hunk 1]
```diff
@@ -250,6 +256,84 @@ size_t DartVM::GetVMLaunchCount() {
...
+  const int sysctl_result =
+      sysctlbyname("kern.maxproc", &maxproc, &maxproc_size, nullptr, 0);
+  if (sysctl_result < 0) {
+    FML_LOG(ERROR)
+        << "Could not execute sysctl() to determine process count limit: "
+        << strerror(errno);
+  }
+
+  const char* warning =
+      "_XXX_ TODO(sjindel): Make a wiki page about this warning. _XXX_";
```
Comment (chinmaygarde) [inline runtime/dart_vm.cc]: You can add the wiki page right away as a stub and link to it here.

Comment (sjindel-google) [reply runtime/dart_vm.cc]: Done.

File: runtime/dart_vm.cc [hunk 2]
```diff
@@ -250,6 +256,84 @@ size_t DartVM::GetVMLaunchCount() {
   return gVMLaunchCount;
 }
 
+// Note: it is __imperative__ that the following two functions are
+// __not__ included in release or profile builds.
+//
+// Please see the following documents for explanation:
+//   - go/decommissioning-dbc
```
[the remark this answers is not in the document]
Comment (sjindel-google) [reply runtime/dart_vm.cc]: Done.

File: runtime/dart_vm.cc [hunk 3]
```diff
@@ -250,6 +256,84 @@ size_t DartVM::GetVMLaunchCount() {
...
+//   - go/decommissioning-dbc-engine
+//   - go/decommissioning-dbc-tools
+#if OS_IOS && (FLUTTER_RUNTIME_MODE == FLUTTER_RUNTIME_MODE_DEBUG)
```
[the remark this answers is not in the document]
Comment (sjindel-google) [reply runtime/dart_vm.cc]: Done.

File: runtime/ptrace_ios.h [hunk 4]
```diff
@@ -0,0 +1,14 @@
+#ifndef FLUTTER_RUNTIME_PTRACE_IOS_H_
```
Comment (chinmaygarde) [inline runtime/ptrace_ios.h]: License block. You can copy the same from a sibling file.

File: runtime/ptrace_ios.cc [hunk 5]
```diff
@@ -0,0 +1,99 @@
+// It is __imperative__ that the functions in this file are __not__ included in
```
Comment (chinmaygarde) [inline runtime/ptrace_ios.cc]: License block. You can copy the same from a sibling file.

File: runtime/ptrace_ios.h (hunk 4 above)
Comment (sjindel-google) [reply runtime/ptrace_ios.h]: Done.

File: runtime/ptrace_ios.cc (hunk 5 above)
Comment (sjindel-google) [reply runtime/ptrace_ios.cc]: Done.

File: runtime/ptrace_ios.h [hunk 6]
```diff
@@ -0,0 +1,18 @@
+// Copyright 2019 The Flutter Authors. All rights reserved.
```
Comment (chinmaygarde) [inline runtime/ptrace_ios.h]: All our licenses have the 2013 copyright date.

Comment (sjindel-google) [reply runtime/ptrace_ios.h]: Done.

**Comment by sjindel-google** [comment]
I've updated the Wiki page: https://github.com/flutter/flutter/wiki/PID-leak-in-iOS-development-workflow
/cc @jonahwilliams @xster

**Comment by sjindel-google** [comment]
The failing checks appear to be due to an infra issue.
I've run the command from the CI (`./flutter/testing/run_tests.sh`) and it passes in all the configurations with "x" above.

**Comment by sjindel-google** [comment]
Rebasing to re-run Fuchsia build test.

**Pull request merged**